### PR TITLE
Introduce statistics counters

### PR DESCRIPTION
**Pull request by ifranzki**
Add support for collecting and displaying mechanism usage statistics for openCryptoki. Usage statistics are collected by openCryptoki on a per user basis. For each user, mechanism usage is counted per configured slot and mechanism. For each mechanism a set of counters exist, one for each cryptographics strength of the cryptographic key used with the mechanism.

### CODE REVIEW

**Comment by ifranzki** [comment]
Rebased to current master, now that PR #487 (Change pkcsslotd to use the generic config file parser) is merged.

File: usr/lib/api/statistics.c [hunk 1]
```diff
@@ -0,0 +1,265 @@
...
+                err = errno;
+                TRACE_ERROR("Failed to create SHM '%s': %s\n",
+                            statistics->shm_name,  strerror(err));
+                OCK_SYSLOG(LOG_ERR, "Failed to create SHM '%s': %s\n",
+                           statistics->shm_name, strerror(err));
+                return CKR_FUNCTION_FAILED;
+            }
+
+            if (fchmod(statistics->shm_handle, S_IRUSR | S_IWUSR) == -1) {
+                err = -errno;
```
Comment (juergenchrist) [inline usr/lib/api/statistics.c:166]: This should be `err = errno` here.

File: usr/lib/api/statistics.c [hunk 2]
```diff
@@ -0,0 +1,265 @@
...
+                TRACE_ERROR("Failed to change mode of SHM '%s': %s\n",
+                            statistics->shm_name,  strerror(err));
+                OCK_SYSLOG(LOG_ERR, "Failed to change mode of SHM '%s': %s\n",
+                           statistics->shm_name, strerror(err));
+                close(statistics->shm_handle);
+                shm_unlink(statistics->shm_name);
+                return CKR_FUNCTION_FAILED;
+            }
+            if (fchown(statistics->shm_handle, user, -1) != 0) {
```
Comment (juergenchrist) [inline usr/lib/api/statistics.c:175]: Why is this needed?  It can only succeed if either `user` is already the owner of the SHM segment, or the current application runs as root.

File: usr/lib/api/statistics.c [hunk 3]
```diff
@@ -0,0 +1,265 @@
...
+        close(statistics->shm_handle);
+        return CKR_FUNCTION_FAILED;
+    }
+
+    /*
+     * If the shared memory segment does not belong to the pkcs11 group or does
+     * not have correct permissions, do not use it.
+     */
+    if (stat_buf.st_uid != (user == -1 ? geteuid() : (uid_t)user) ||
+        stat_buf.st_gid != (user == -1 ? geteuid() : (uid_t)user) ||
```
Comment (juergenchrist) [inline usr/lib/api/statistics.c:210]: Comparing group id against user id here.  Should this be the id of the pkcs11 group?  Also I do not see the code that chowns the newly created segment to the pkcs11 group.

File: usr/lib/icsf_stdll/pbkdf.c [hunk 4]
```diff
@@ -371,10 +410,28 @@ CK_RV pbkdf(CK_BYTE * password, CK_ULONG len, CK_BYTE * salt, CK_BYTE * dkey,
...
+    if (rc == CKR_OK && tokdata != NULL &&
+        (tokdata->statistics->flags & STATISTICS_FLAG_COUNT_INTERNAL) != 0) {
+        tokdata->statistics->increment_func(tokdata->statistics,
+                                            tokdata->slot_id, &mech,
+                                            POLICY_STRENGTH_IDX_0);
+        if ((tokdata->statistics->flags & STATISTICS_FLAG_COUNT_IMPLICIT) != 0) {
+            /*
+             * We use CKM_PKCS5_PBKD2 with CKP_PKCS5_PBKD2_HMAC_SHA256.
+             * Use strength 0 because the HMAC key is the pin, and it is max
+             * 8 char (i.e. 64 bit) long, which is way below 112 bit anyway.
```
Comment (juergenchrist) [inline usr/lib/icsf_stdll/pbkdf.c:422]: Technically that would be up to whoever configures the strength.conf-file.

File: usr/sbin/pkcsslotd/slotmgr.c [hunk 5]
```diff
@@ -465,12 +465,50 @@ static int config_parse_slot(const char *config_file,
...
+{
+    struct ConfigBaseNode *c;
+    int i;
+
+    confignode_foreach(c, statistics->value, i) {
+        DbgLog(DL3, "Config node: '%s' type: %u line: %u\n",
+               c->key, c->type, c->line);
+
+        if (c->type == CT_BARE && strcmp(c->key, "off") == 0) {
+            socketData.flags &= ~(FLAG_STATISTICS_ENABLED |
```
Comment (juergenchrist) [inline usr/sbin/pkcsslotd/slotmgr.c:479]: That makes order somewhat important. `off, implicit` leads to different flags than `implicit, off`.  Did not check the remaining code if this might be a problem.  Personally I would not change `socketData.flags` in the off case directly, but remember to change them before returning and remove all flags at the end.

File: configure.ac [hunk 6]
```diff
@@ -681,6 +690,7 @@ AC_CONFIG_FILES([Makefile				\
 		 man/man1/pkcsep11_migrate.1		\
 		 man/man1/pkcsep11_session.1		\
 		 man/man1/pkcstok_migrate.1		\
+		  man/man1/pkcsstats.1			\
```
Comment (juergenchrist) [inline configure.ac:693]: Just my personal opinion, but we provide configure-switches for all tools (why?), but unconditionally build the man-pages for all tools.  That seems weird to me.

File: man/man1/pkcsstats.1.in [hunk 7]
```diff
@@ -0,0 +1,135 @@
...
+The strength configuration file defines how the strength is determined for the
+various key types. A strength of zero is used to count those mechanisms that
+do not use a key, or where the key strength is less than 112 bits.
+.PP
+.B Note:
+The strength does not specify the cryptographic strength of the mechanism, but
+the cryptographic strength of the key used with the mechanism (if any).
+For example, usage of mechanism CKM\_SHA256 is reported under strength 0,
+because no key is used with this mechanism. However, usage of mechanism
+CKM\_AES\_CBC is reported under strength 128, 192, or 256, dependent on the
```
Comment (juergenchrist) [inline man/man1/pkcsstats.1.in:38]: Technically, if a strength.json maps all symmetric keys to strength 0 (don't know why someone would do that, but it is possible), CKM_AES_CBC would end up in strength 0 bucket.

File: man/man1/pkcsstats.1.in [hunk 8]
```diff
@@ -0,0 +1,135 @@
...
+in the openCryptoki configuration file
+\fB/etc/opencryptoki/opencryptoki.conf\fP.
+By default only explicit mechanism usage statistics from PKCS#11 applications
+are collected.
+.PP
+Optionally, implicit mechanism usage statistics can be collected, where
+additional mechanisms are specified in mechanism parameters. For example,
+RSA\-PSS or RSA\-OAEP allow to specify a hash mechanism and a mask generation
+function (MGF) in the mechanism parameter. ECDH allows to specify a key
+derivation function (KDF) in the mechanism parameter.
```
Comment (juergenchrist) [inline man/man1/pkcsstats.1.in:52]: You do not mention the PRFs used by PBKDF2, but from the code I see that they are counted, too.  Should we mention them here, too?

File: man/man1/pkcsstats.1.in [hunk 9]
```diff
@@ -0,0 +1,135 @@
...
+.PP
+Also optionally, opencryptoki\-internal mechanism usage statistics can be
+collected. This collects usage statistics for crypto operations used internally
+for pin handling and encryption of private token objects in the data store.
+.PP
+.B Note:
+Implicit or internal mechanism usage can not be distinguished from explicit
+mechanism usage of PKCS#11 applications in the displayed statistics.
+.PP
+Statistics are collected in a shared memory segment per user. This shared memory
```
Comment (juergenchrist) [inline man/man1/pkcsstats.1.in:62]: Can we somehow make clear that we are talking about a POSIX shared memory segment and not a SysV IPC shared memory segment?  Might be of interest for people dealing with namespaces.

File: man/man1/pkcsstats.1.in [hunk 10]
```diff
@@ -0,0 +1,135 @@
...
+(when not already existent). The shared memory segmenst can be deleted using
+the \fBpkcsstats\fP command with the \fB\-\-delete\fP, or \fB\-\-delete\-all\fP
+options.
+.PP
+The usage of a mechanism is counted once when the cryptographic operation is
+sucessfully initialized, i.e. during \fBC_DigestInit\fP, \fBC_EncryptInit\fP,
+\fBC_DecryptInit\fP, \fBC_SignInit\fP, \fBC_SignRecoverInit\fP, and
+\fBC_VerifyInit\fP. Multi-part operations involving the update functions like
+\fBC_DigestUpdate\fP, \fBC_EncryptUpdate\fP, \fBC_DecryptUpdate\fP,
+\fBC_SignUpdate\fP, and \fBC_VerifyUpdate\fP, are not counted additionally.
```
Comment (juergenchrist) [inline man/man1/pkcsstats.1.in:77]: Should we mention key-gen and key-derive?

File: usr/sbin/pkcsstats/pkcsstats.mk [hunk 11]
```diff
@@ -0,0 +1,16 @@
+sbin_PROGRAMS += usr/sbin/pkcsstats/pkcsstats
+
+usr_sbin_pkcsstats_pkcsstats_LDFLAGS = -lcrypto -ldl -lrt
+
+usr_sbin_pkcsstats_pkcsstats_CFLAGS  =			\
+	-DOCK_TOOL					\
+	-DSTDLL_NAME=\"pkcsstats\"			\
+	-I${srcdir}/usr/include 			\
+	-I${srcdir}/usr/lib/common 			\
+	-I${srcdir}/usr/lib/api
```
Comment (juergenchrist) [inline usr/sbin/pkcsstats/pkcsstats.mk:10]: You also need `-I${top_builddir}/usr/lib/api` since this is where `mechtable-gen.h` is placed.

File: usr/sbin/pkcsstats/pkcsstats.c [hunk 12]
```diff
@@ -0,0 +1,867 @@
...
+
+    return rc;
+}
+
+int init_ock(void **dll, CK_FUNCTION_LIST_PTR *func_list)
+{
+    void (*sym_ptr)();
+    CK_RV rc;
+
+    *dll = dlopen("libopencryptoki.so", RTLD_NOW);
```
Comment (juergenchrist) [inline usr/sbin/pkcsstats/pkcsstats.c:640]: What is the benefit of using library loading here?  We could just directly link against libopencryptoki.so and get rid of all the `dlsym` and the like...

File: Makefile.am [hunk 13]
```diff
@@ -235,10 +236,12 @@ ci-installcheck: ci-prepare installcheck
 	killall -HUP pkcsslotd || true
 	@sbindir@/pkcsslotd
 	cd ${srcdir}/testcases && export PKCSLIB=@libdir@/opencryptoki/libopencryptoki.so && export PKCS11_USER_PIN=$(PKCS11_USER_PIN) && ./misc_tests/p11sak_test.sh | tee log-p11sak.txt
+	@sbindir@/pkcsstats
```
Comment (juergenchrist) [inline Makefile.am:239]: Doesn't this only print the statistics for the current user?  The CI runs the tests under su to a special user `tester`.  Don't we then need to add the `--all` flag?

File: Makefile.am [hunk 14]
```diff
@@ -235,10 +236,12 @@ ci-installcheck: ci-prepare installcheck
...
 	killall -HUP pkcsslotd
 	@echo "done"
 
 ci-cleanup:
+	@sbindir@/pkcsstats --delete || true
```
Comment (juergenchrist) [inline Makefile.am:244]: Same here?

File: usr/lib/api/statistics.c (hunk 1 above)
Comment (ifranzki) [reply usr/lib/api/statistics.c:166]: fixed

File: usr/lib/api/statistics.c (hunk 2 above)
Comment (ifranzki) [reply usr/lib/api/statistics.c:175]: yes, you are right, will remove the fchown block.

File: usr/lib/api/statistics.c (hunk 3 above)
Comment (ifranzki) [reply usr/lib/api/statistics.c:210]: oups,... . no pkcs11 group used here intentionally. the segment is supposed to belong only to the user.
Will change to only check st_uid and st_mode, but not st_gid

File: usr/lib/icsf_stdll/pbkdf.c (hunk 4 above)
Comment (ifranzki) [reply usr/lib/icsf_stdll/pbkdf.c:422]: I think we already discussed this offline.... The conclusion was that it makes absolutely no sense to map this to anything higher than or equal to 112 bits strength (although it is technically possible to do so). So we simply assume below 112 bits and thus always use strength zero.

File: usr/sbin/pkcsslotd/slotmgr.c (hunk 5 above)
Comment (ifranzki) [reply usr/sbin/pkcsslotd/slotmgr.c:479]: Well, yes, but if flag FLAG_STATISTICS_ENABLED is off, then the other flags are not checked at all.  So if anywhere 'off' is specified, the FLAG_STATISTICS_ENABLED  is off and thus statistics are disabled, no matter if and how often the other keywords are specified. 

One could also specify something like `on, implicit, off, internal, on`. To be 100% correct we should also check that each keyword appears only once, and only the right combinations and order.  Not sure if it is worth to be that stricht though....

File: man/man1/pkcsstats.1.in (hunk 7 above)
Comment (ifranzki) [reply man/man1/pkcsstats.1.in:38]: Right, this was mean with the part in the bracket: 
_...dependent on the cryptographic size of the AES key used with it (**and the definitions in the strength configuration file**)_

File: man/man1/pkcsstats.1.in (hunk 8 above)
Comment (ifranzki) [reply man/man1/pkcsstats.1.in:52]: good point. Currently the PBKDF2 mech is only used for OCK-internal crypto, but we might want to add support for this mechanism to tokens at a alter time. So I'll add it to the description in the man page.

File: man/man1/pkcsstats.1.in (hunk 9 above)
Comment (ifranzki) [reply man/man1/pkcsstats.1.in:62]: good point. changed.

File: man/man1/pkcsstats.1.in (hunk 10 above)
Comment (ifranzki) [reply man/man1/pkcsstats.1.in:77]: Well, this section talks about the Init calls versus Udpate/Final and tries to make clear that only the init call is counted. For Keygen/Derive/Wrap, there is no such thing as a Init/Update/Final call. Its just one call to generate, derive, wrap a key. 
But maybe it is worth to add a section for those other calls (like KeyGen, Derive, Wrap) to make clear that they are also counted.

File: usr/sbin/pkcsstats/pkcsstats.mk (hunk 11 above)
Comment (ifranzki) [reply usr/sbin/pkcsstats/pkcsstats.mk:10]: good catch!

File: usr/sbin/pkcsstats/pkcsstats.c (hunk 12 above)
Comment (ifranzki) [reply usr/sbin/pkcsstats/pkcsstats.c:640]: well, that's the way all our tools do it....

We could certainly change that at some point in time, but I would not want to do this in this PR. Will certainly need changes in the make files to ensure that the library is build before the tool.

File: Makefile.am (hunk 13 above)
Comment (ifranzki) [reply Makefile.am:239]: good catch. Yes, lets use --all (and --delete-all and --reset-all at the other places).

File: configure.ac (hunk 6 above)
Comment (ifranzki) [reply configure.ac:693]: Yes that's somewhat inconsistent. But at least we only include the built man page according to the configure-switch in man1.mk....

File: Makefile.am (hunk 14 above)
Comment (ifranzki) [reply Makefile.am]: fixed

File: usr/lib/common/loadsave.c [hunk 15]
```diff
@@ -1749,18 +1816,37 @@ static CK_RV aes_256_unwrap(unsigned char key[32],
...
 {
+    CK_RV rc;
+
     if (tokdata->version < TOK_NEW_DATA_STORE)
         return generate_master_key_old(tokdata, key);
 
     /* generate a 256-bit AES key */
-    return rng_generate(tokdata, key, 32);
+    rc = rng_generate(tokdata, key, 32);
+
```
Comment (jschmidb) [inline usr/lib/common/loadsave.c:1840]: When you are at it: there are constants in common/defs.h for the various key lengths:
```
#define AES_KEY_SIZE_256 32
#define AES_KEY_SIZE_192 24
#define AES_KEY_SIZE_128 16
```
So we could use AES_KEY_SIZE_256 here, instead of 32. But I see, this was old code.

File: man/man5/opencryptoki.conf.5.in [hunk 16]
```diff
@@ -18,6 +18,29 @@ The following global definitions are valid:
...
+In addition to enabling statistics collection for mechanisms used by PKCS#11
+applications, you can specify \fB(on,implicit)\fP to also enable collection
+of implicit mechanism usage, where additional mechanisms are specified in
+mechanism parameters. For example, RSA-PSS or RSA-OAEP allow to specify a hash
+mechanism and a mask generation function (MGF) in the mechanism parameter.
+ECDH allows to specify a key derivation function (KDF) in the mechanism
+parameter.
+
+You can additionaly enable statistics collection of mechanisms internally used
+by Opencryptoki by specifying \fB(on,internal)\fP. This additionally collects
```
Comment (jschmidb) [inline man/man5/opencryptoki.conf.5.in:37]: additionally

File: man/man1/pkcsstats.1.in [hunk 17]
```diff
@@ -0,0 +1,141 @@
...
+.B pkcsstats
+.BR \-\-help | \-h
+.br
+
+.SH DESCRIPTION
+Displays mechanism usage statistics for openCryptoki. Usage statistics are
+collected by openCryptoki on a per user basis. For each user, mechanism
+usage is counted per configured slot and mechanism. For each mechanism a set of
+counters exist, one for each cryptographics strength of the cryptographic
+key used with the mechanism.
```
Comment (jschmidb) [inline man/man1/pkcsstats.1.in:24]: ... each cryptographic strength ...

File: man/man1/pkcsstats.1.in [hunk 18]
```diff
@@ -0,0 +1,141 @@
...
+\fB/etc/opencryptoki/opencryptoki.conf\fP.
+By default only explicit mechanism usage statistics from PKCS#11 applications
+are collected.
+.PP
+Optionally, implicit mechanism usage statistics can be collected, where
+additional mechanisms are specified in mechanism parameters. For example,
+RSA\-PSS or RSA\-OAEP allow to specify a hash mechanism and a mask generation
+function (MGF) in the mechanism parameter. ECDH allows to specify a key
+derivation function (KDF) in the mechanism parameter. The PBKDF2 mechanism
+allow to specify a pseudo random function (PRF) in the mechanism parameter.
```
Comment (jschmidb) [inline man/man1/pkcsstats.1.in:53]: allows to specify ...

File: man/man1/pkcsstats.1.in [hunk 19]
```diff
@@ -0,0 +1,141 @@
...
+.PP
+.B Note:
+Implicit or internal mechanism usage can not be distinguished from explicit
+mechanism usage of PKCS#11 applications in the displayed statistics.
+.PP
+Statistics are collected in a POSIX shared memory segment per user. This shared
+memory segment contains all counters for all configured slots, mechanisms, and
+strengths. The shared memory segments are named
+\fBvar.lib.opencryptoki_stats_<uid>\fP, where \fBuid\fP is the numeric user\-id
+of the user the statistics belong to. The shared memory segmenst are
```
Comment (jschmidb) [inline man/man1/pkcsstats.1.in:67]: segments

File: man/man1/pkcsstats.1.in [hunk 20]
```diff
@@ -0,0 +1,141 @@
...
+automatically created for a user on the first attempt to colletc statistics
+(when not already existent). The shared memory segmenst can be deleted using
```
Comment (jschmidb) [inline man/man1/pkcsstats.1.in:69]: to collect ...   and again segmenst -> segments

File: man/man1/pkcsstats.1.in [hunk 21]
```diff
@@ -0,0 +1,141 @@
...
+Shows the accumulated statistics from all users. Only the \fBroot\fP user can
+display the accumulated statistics from other users.
+.TP
+.BR \-A ", " \-\-all
+Shows the statistics from all users. Only the \fBroot\fP user can display
+statistics from all users.
+.TP
+.BR \-a ", " \-\-all\-mechs
+Shows the statistics for all mechanisms, also those with all\-zero counters.
+If this option is omitted, only those mechaniusms are displayed where at least
```
Comment (jschmidb) [inline man/man1/pkcsstats.1.in:104]: mechanisms

File: usr/sbin/pkcsstats/pkcsstats.c [hunk 22]
```diff
@@ -0,0 +1,867 @@
...
+    rc = open_shm(user_id, user_name, num_slots, &shm_fd, &shm_data, &shm_size);
+    if (rc != 0)
+        return rc;
+
+    rc = for_all_slots(reset_slot_cb, NULL, shm_data, shm_size,
+                       num_slots, slots, slot_id_specified, slot_id);
+
+    if (rc == 0) {
+        if (slot_id_specified)
+            printf("Reseted statistics for user '%s' and slot %lu\n",
```
Comment (jschmidb) [inline usr/sbin/pkcsstats/pkcsstats.c:320]: Resetted

File: usr/sbin/pkcsstats/pkcsstats.c [hunk 23]
```diff
@@ -0,0 +1,867 @@
...
+                   user_name, slot_id);
+        else
+            printf("Reseted statistics for user '%s'\n", user_name);
```
Comment (jschmidb) [inline usr/sbin/pkcsstats/pkcsstats.c:323]: Resetted

File: usr/sbin/pkcsstats/pkcsstats.c [hunk 24]
```diff
@@ -0,0 +1,867 @@
...
+        {"delete", no_argument, NULL, 'd'},
+        {"delete-all", no_argument, NULL, 'D'},
+        {"help", no_argument, NULL, 'h'},
+        {0, 0, 0, 0}
+    };
+
+    while ((opt = getopt_long(argc, argv, "U:SAas:rRdDvh", long_opts, NULL)) != -1) {
+        switch (opt) {
+        case 'U':
+            if((pswd = getpwnam(optarg)) == NULL){
```
Comment (jschmidb) [inline usr/sbin/pkcsstats/pkcsstats.c:712]: space after if? Also below ... ?

File: usr/sbin/pkcsstats/pkcsstats.c [hunk 25]
```diff
@@ -0,0 +1,867 @@
...
+                warnx("The username '%s' is not known on this system.", optarg);
```
Comment (jschmidb) [inline usr/sbin/pkcsstats/pkcsstats.c:713]: All other messages don't have a dot at the end ....

File: man/man5/opencryptoki.conf.5.in (hunk 16 above)
Comment (ifranzki) [reply man/man5/opencryptoki.conf.5.in:37]: fixed

File: man/man1/pkcsstats.1.in (hunk 17 above)
Comment (ifranzki) [reply man/man1/pkcsstats.1.in:24]: fixed

File: man/man1/pkcsstats.1.in (hunk 18 above)
Comment (ifranzki) [reply man/man1/pkcsstats.1.in:53]: fixed

File: man/man1/pkcsstats.1.in (hunk 19 above)
Comment (ifranzki) [reply man/man1/pkcsstats.1.in:67]: fixed

File: man/man1/pkcsstats.1.in (hunk 20 above)
Comment (ifranzki) [reply man/man1/pkcsstats.1.in:69]: fixed

File: man/man1/pkcsstats.1.in (hunk 21 above)
Comment (ifranzki) [reply man/man1/pkcsstats.1.in:104]: fixed

File: usr/sbin/pkcsstats/pkcsstats.c (hunk 22 above)
Comment (ifranzki) [reply usr/sbin/pkcsstats/pkcsstats.c:320]: fixed

File: usr/sbin/pkcsstats/pkcsstats.c (hunk 23 above)
Comment (ifranzki) [reply usr/sbin/pkcsstats/pkcsstats.c:323]: fixed

File: usr/sbin/pkcsstats/pkcsstats.c (hunk 24 above)
Comment (ifranzki) [reply usr/sbin/pkcsstats/pkcsstats.c:712]: fixed

File: usr/sbin/pkcsstats/pkcsstats.c (hunk 25 above)
Comment (ifranzki) [reply usr/sbin/pkcsstats/pkcsstats.c:713]: fixed

File: usr/lib/api/statistics.h [hunk 26]
```diff
@@ -0,0 +1,66 @@
...
+#include <pkcs11types.h>
+#include "slotmgr.h"
+#include "mechtable.h"
+#include "supportedstrengths.h"
+
+/*
+ * Statistics are collected in a shared memory segment per user.
+ * The statistics shared memory segment has the following layout:
+ * - For each configured slot:
+ *    - For each mechanisms supported:
```
Comment (jschmidb) [inline usr/lib/api/statistics.h:22]: For each supported mechanism (?)

File: usr/lib/api/statistics.h [hunk 27]
```diff
@@ -0,0 +1,66 @@
...
+ *       - one counter (counter_t) for non-key mechanisms (strength=0)
+ *       - for each supported strength one counter (counter_t each)
+ *
```
Comment (jschmidb) [inline usr/lib/api/statistics.h:25]: one counter for each supported strength (?)

File: usr/lib/api/statistics.h (hunk 26 above)
Comment (ifranzki) [reply usr/lib/api/statistics.h:22]: fixed

File: usr/lib/api/statistics.h (hunk 27 above)
Comment (ifranzki) [reply usr/lib/api/statistics.h:25]: fixed

**Comment by ifranzki** [comment]
Added one more commit to add machine readable JSON format output.

File: usr/sbin/pkcsstats/pkcsstats.c [hunk 28]
```diff
@@ -406,24 +407,45 @@ int get_token_infos(CK_FUNCTION_LIST_PTR func_list, CK_SLOT_ID slot,
...
     UNUSED(mech);
-    UNUSED(private);
     UNUSED(ofs);
 
-    printf("%-30s |", mech_name);
+    if (dm->json && dm->first_mech == false)
+        printf(",");
+
+    if (dm->json)
+        printf("\n\t\t\t\t\t\t{\n\t\t\t\t\t\t\t\"mechanism\": \"%s\",\n", mech_name);
```
Comment (juergenchrist) [inline usr/sbin/pkcsstats/pkcsstats.c:430]: This is supposed to be the machine-readable output, right?  If yes, why do we care so much about indentation here?  We could just produce minimal JSON (maybe not minimal since I am not opposed to new lines) instead of this vast amount of tabs...

File: usr/sbin/pkcsstats/pkcsstats.c [hunk 29]
```diff
@@ -837,12 +918,17 @@ int main(int argc, char **argv)
         goto done;
     }
 
+    if (json)
+        printf("{\n\t\"users\": [\n");
```
Comment (juergenchrist) [inline usr/sbin/pkcsstats/pkcsstats.c:922]: Why a struct containing only one property?  We could simply print a json array here?

**Review by juergenchrist — APPROVED** [review]

File: usr/sbin/pkcsstats/pkcsstats.c (hunk 28 above)
Comment (ifranzki) [reply usr/sbin/pkcsstats/pkcsstats.c:430]: Well, yes, but I also wanted it to be human readable to some extend....
I could indent with blanks (i.e. 2 blanks) instead of tabs, but that would produce more data than tabs....

File: usr/sbin/pkcsstats/pkcsstats.c (hunk 29 above)
Comment (ifranzki) [reply usr/sbin/pkcsstats/pkcsstats.c:922]: Thats the root of the JSON tree, isn't the root required to be a struct ?

Comment (ifranzki) [reply usr/sbin/pkcsstats/pkcsstats.c:922]: and I wanted to call that think 'users'. So just having an array would not tell what the array is for.

**Review by jschmidb — APPROVED** [review]

File: usr/sbin/pkcsstats/pkcsstats.c (hunk 29 above)
Comment (juergenchrist) [reply usr/sbin/pkcsstats/pkcsstats.c:922]: > isn't the root required to be a struct ?

No, any value is fine according to new ECMA standard.  So specifically an array would be okay.

Comment (juergenchrist) [reply usr/sbin/pkcsstats/pkcsstats.c:922]: > I wanted to call that think 'users'.

Fine with me.  If this is supposed to be consumed by other tools you anyway need to clearly describe this API and consider it an API.  I am just not sure which other top-level key you might want to add in the future...

Comment (ifranzki) [reply usr/sbin/pkcsstats/pkcsstats.c:922]: The `zcryptstats` tool of s390-tools also provides JSON output, and it adds system information and a timestamp into the top level node. 
So maybe we should do this here, too ? Opinions?

```
static int json_print_header(void)
{
	char timestamp[64];
	struct utsname un;
	struct tm *tm;
	time_t t;

	time(&t);
	tm = localtime(&t);
	strftime(timestamp, sizeof(timestamp), "%x", tm);

	if (uname(&un) != 0)
		return -errno;

	printf("{\"zcryptstats\": {\n");
	printf("\t\"host\": {\n");
	printf("\t\t\"nodename\": \"%s\",\n", un.nodename);
	printf("\t\t\"sysname\": \"%s\",\n", un.sysname);
	printf("\t\t\"release\": \"%s\",\n", un.release);
	printf("\t\t\"machine\": \"%s\",\n", un.machine);
	printf("\t\t\"date\": \"%s\",\n", timestamp);
	printf("\t\t\"statistics\": [\n");

	return 0;
}
```

Comment (juergenchrist) [reply usr/sbin/pkcsstats/pkcsstats.c:922]: Might be a good idea to already include it.  Otherwise people will be forced to write some wrapper exporters to modify the JSON to add this data.

But please no `%x` for `strftime`.  Locale-awareness might be a good idea for user-readable things, but is a horror for something I want to process further via some code.  Personally I would go for `%F` which produces some ISO-standard date format.

Comment (ifranzki) [reply usr/sbin/pkcsstats/pkcsstats.c]: done

File: usr/lib/api/statistics.c (hunk 2 above)
Comment (ifranzki) [reply usr/lib/api/statistics.c]: Maybe add this back in? This function is actually never called for another user, but if it would, then we should do the fchown for that user I guess.

Comment (ifranzki) [reply usr/lib/api/statistics.c]: @juergenchrist Do you want to comment on this? 
Other than that I would say that this PR is ready to merge.

Comment (juergenchrist) [reply usr/lib/api/statistics.c]: Currently I see this function as `static` with exactly one caller who passes `-1` as user.  In general: what should be accomplished with the `fchown`?  Just sounds weird to me that one user might create the SHM-User-Segment of a different user.  The combination "read access" and "different user" makes sense, but then the segment should not be created.

**Review by p-steuer — APPROVED** [review]

File: usr/lib/api/statistics.c (hunk 2 above)
Comment (ifranzki) [reply usr/lib/api/statistics.c]: The original idea was that this function could be called by pkcsstats too open the segment of any user (when running as root), but it turned of that linking this module with the tool would cause too much trouble (many other unresolved externals), so it is only used at once place at the moment.  So yes, lets keep it as is.
